File: rpc/clnt/clnt.go
// Package rpcclnt is the client stub that (un)marshals argument
// protobufs and result protobufs.  By default it uses sigmaos to
// send/receive the request and receive reply.  But caller can pass
// into its own RPCCh interface to replace the default (e.g., to
// send/receive over a pipe).

package clnt

import (
	"fmt"
	"time"

	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/timestamppb"

	db "sigmaos/debug"
	spproxyproto "sigmaos/proxy/sigmap/proto"
	"sigmaos/rpc"
	"sigmaos/rpc/clnt/channel"
	"sigmaos/rpc/clnt/delegation"
	rpcclntopts "sigmaos/rpc/clnt/opts"
	rpcproto "sigmaos/rpc/proto"
	"sigmaos/serr"
	sessp "sigmaos/session/proto"
	"sigmaos/shmem"
	sp "sigmaos/sigmap"
	"sigmaos/util/perf"
)

type RPCcall struct {
	method string
	iov    sessp.IoVec
}

type RPCClnt struct {
	si                    *rpc.StatInfo
	ch                    channel.RPCChannel
	delegatedRPCCh        channel.RPCChannel
	rc                    *delegation.ReplyCache
	useShmemDelegatedRPCs bool
	shmemSegment          *shmem.Segment
}

// XXX TODO Shouldn't take pn here
func NewRPCClnt(pn string, opts ...*rpcclntopts.RPCClntOption) (*RPCClnt, error) {
	s := time.Now()
	defer func() {
		db.DPrintf(db.ATTACH_LAT, "NewSigmaRPCClnt %q lat %v", pn, time.Since(s))
	}()

	rpcOpts := rpcclntopts.NewEmptyRPCClntOptions()
	for _, opt := range opts {
		opt.Apply(rpcOpts)
	}

	ch, err := rpcOpts.NewRPCChannel(pn)
	if err != nil {
		return nil, err
	}
	delCh, err := rpcOpts.NewDelegatedRPCChannel(sp.NOT_SET)
	if err != nil {
		return nil, err
	}
	return &RPCClnt{
		si:                    rpc.NewStatInfo(),
		ch:                    ch,
		delegatedRPCCh:        delCh,
		rc:                    delegation.NewReplyCache(),
		useShmemDelegatedRPCs: rpcOpts.UseShmemDelegatedRPCs,
		shmemSegment:          rpcOpts.ShmemSegment,
	}, nil
}

func WrapRPCRequest(method string, arg proto.Message) (*sessp.IoVec, error) {
	iniov := sessp.NewUnallocatedIoVec(0, nil)
	inblob := rpc.GetBlob(arg)
	if inblob != nil {
		iniov = inblob.GetIoVec()
		inblob.ClearIoVec()
	}
	argBytes, err := proto.Marshal(arg)
	if err != nil {
		return nil, err
	}
	iniov.InsertFrame(0, sessp.NewFrame(argBytes, nil))
	return WrapMarshaledRPCRequest(method, iniov)
}

func WrapMarshaledRPCRequest(method string, iniov *sessp.IoVec) (*sessp.IoVec, error) {
	req := rpcproto.Req{Method: method}
	wrapperBytes, err := proto.Marshal(&req)
	if err != nil {
		return nil, serr.NewErrError(err)
	}
	iniov.InsertFrame(0, sessp.NewFrame(wrapperBytes, nil))
	return iniov, nil
}

func (rpcc *RPCClnt) runWrappedRPC(delegate bool, method string, iniov *sessp.IoVec, outiov *sessp.IoVec) error {
	var err error
	start := time.Now()
	if delegate {
		// Sanity check
		if rpcc.delegatedRPCCh == nil {
			db.DFatalf("Try to run delegated RPC with unset delegated RPC channel")
		}
		err = rpcc.delegatedRPCCh.SendReceive(iniov, outiov)
	} else {
		err = rpcc.ch.SendReceive(iniov, outiov)
	}
	if err != nil {
		return err
	}
	// Record stats
	rpcc.si.Stat(method, time.Since(start).Microseconds())
	return nil
}

func processWrappedRPCRep(outiov *sessp.IoVec, res proto.Message, outblob *rpcproto.Blob) error {
	if outiov.Len() == 0 {
		return serr.NewErr(serr.TErrUnreachable, "empty RPC reply")
	}
	rep := &rpcproto.Rep{}
	if err := proto.Unmarshal(outiov.GetFrame(0).GetBuf(), rep); err != nil {
		return serr.NewErrError(err)
	}
	if rep.Err.ErrCode != 0 {
		return sp.NewErr(rep.Err)
	}
	// A reply that carries no error must carry the result message (and its blob
	// frames, if any). Fewer frames than that means the reply was truncated;
	// report it rather than indexing past the end.
	if outiov.Len() < 2 {
		return serr.NewErr(serr.TErrUnreachable, fmt.Sprintf("short RPC reply: %d frames, no error", outiov.Len()))
	}
	if err := proto.Unmarshal(outiov.GetFrame(1).GetBuf(), res); err != nil {
		return err
	}
	if outblob != nil {
		// Need to get the blob again, because its value will be reset during
		// unmarshaling
		outblob = rpc.GetBlob(res)
		// Set the IoVec to handle replies with blobs
		outblob.SetIoVecBufs(outiov.ToByteSlices()[2:])
	}
	return nil
}

// RPC handles arg and res that contain a Blob specially: it removes
// the blob from the message and pass it down in an IoVec to avoid
// marshaling overhead of large blobs.
func (rpcc *RPCClnt) RPC(method string, arg proto.Message, res proto.Message) error {
	return rpcc.rpc(false, method, arg, res)
}

func (rpcc *RPCClnt) rpc(delegate bool, method string, arg proto.Message, res proto.Message) error {
	iniov, err := WrapRPCRequest(method, arg)
	if err != nil {
		return err
	}
	// Prepend 2 empty slots to the out iovec: one for the rpcproto.Rep
	// wrapper, and one for the marshaled res proto.Message
	outiov := sessp.NewUnallocatedIoVec(2, nil)
	outblob := rpc.GetBlob(res)
	if outblob != nil { // handle blob
		// Get the reply's blob, if it has one, so that data can be read directly
		// into buffers in its IoVec
		outiov.AppendFrames(outblob.GetIoVec().GetFrames())
	}
	if err := rpcc.runWrappedRPC(delegate, method, iniov, outiov); err != nil {
		return err
	}
	if err := processWrappedRPCRep(outiov, res, outblob); err != nil {
		return err
	}
	return nil
}

// OutgoingDelegatedRPC sends an RPC to the boot script via the SPProxySrv
func (rpcc *RPCClnt) OutgoingDelegatedRPC(rpcIdx uint64, method string, arg proto.Message) error {
	iniov, err := WrapRPCRequest(method, arg)
	if err != nil {
		return err
	}
	blob := &rpcproto.Blob{}
	blob.SetIoVec(iniov)
	req := &spproxyproto.SigmaOutgoingDelegatedRPCReq{
		RPCIdx:          rpcIdx,
		Blob:            blob,
		TransferStartPB: timestamppb.New(time.Now()),
	}
	rep := &spproxyproto.SigmaErrRep{}
	if err := rpcc.rpc(true, "SPProxySrvAPI.OutgoingDelegatedRPC", req, rep); err != nil {
		db.DPrintf(db.ERROR, "Err OutgoingDelegatedRPC(%v): %v", rpcIdx, err)
		return err
	}
	if rep.Err.ErrCode != 0 {
		return sp.NewErr(rep.Err)
	}
	return nil
}

// DelegatedRPC handles a delegated RPC (requesting the response from
// SPProxySrv via the delegated RPC channel), retreiving a res that contains a
// Blob specially: it removes the blob from the message and pass it down in an
// IoVec to avoid marshaling overhead of large blobs.
func (rpcc *RPCClnt) DelegatedRPC(rpcIdx uint64, res proto.Message) (time.Duration, error) {
	var nIOVec int
	if !rpcc.useShmemDelegatedRPCs {
		// Prepend 2 empty slots to the out iovec: one for the rpcproto.Rep
		// wrapper, and one for the marshaled res proto.Message
		nIOVec = 2
	} else {
		// Prepend nothing to out IOV. Everything will be passed in shared memory
	}
	outiov := sessp.NewUnallocatedIoVec(nIOVec, nil)
	outblob := rpc.GetBlob(res)
	if !rpcc.useShmemDelegatedRPCs {
		// If not using shared memory, append blob IOVecs to the out IOVec
		if outblob != nil { // handle blob
			// Get the reply's blob, if it has one, so that data can be read directly
			// into buffers in its IoVec
			outiov.AppendFrames(outblob.GetIoVec().GetFrames())
		}
	}
	req := &spproxyproto.SigmaDelegatedRPCReq{
		RPCIdx:   rpcIdx,
		UseShmem: rpcc.useShmemDelegatedRPCs,
	}
	rep := &spproxyproto.SigmaDelegatedRPCRep{
		Blob: &rpcproto.Blob{
			Iov: outiov.ToByteSlices(),
		},
	}
	start := time.Now()
	// Check if the reply has already been cached client-side
	err, ok := rpcc.rc.Get(rpcIdx, rep)
	if !ok {
		rpcc.rc.Register(rpcIdx)
		// If delegated RPC reply wasn't cached on the client-side, request it from
		// the SPProxy
		err = rpcc.rpc(true, "SPProxySrvAPI.GetDelegatedRPCReply", req, rep)
		rpcc.rc.Put(rpcIdx, rep, err)
	} else {
		// If delegated RPC reply was cached on the client-side
		db.DPrintf(db.RPCCLNT, "Get DelegatedRPC(%v) cached", rpcIdx)
	}
	if err != nil {
		return 0, err
	}
	// Before touching the reply's frames: the delegated RPC itself may have
	// failed, in which case there is no result message or blob to map — only the
	// error. Checking after the frame handling below meant an errored delegated
	// RPC underflowed the frame count to -1 and died on the sanity check with a
	// message about buffer counts, hiding the error that actually happened.
	if rep.Err.ErrCode != 0 {
		return 0, sp.NewErr(rep.Err)
	}
	if rpcc.useShmemDelegatedRPCs {
		// Set IOVec from shared memory region
		if rep.UseShmem {
			// The reply is the wrapper, the result message, and one frame per
			// blob buffer the caller supplied. Fewer frames than that (with no
			// error reported above) means a truncated reply: report it rather
			// than mapping frames that aren't there.
			if len(rep.ShmOffs) < 2 {
				return 0, serr.NewErr(serr.TErrUnreachable, fmt.Sprintf("short shmem delegated RPC(%v) reply: %d frames", rpcIdx, len(rep.ShmOffs)))
			}
			if len(outblob.Iov) != len(rep.ShmOffs)-2 {
				db.DFatalf("Wrong number of buffers supplied for shared-memory delegated RPC: %v != %v", len(outblob.Iov), len(rep.ShmOffs)-2)
			}
			frames := make([]*sessp.Tframe, len(rep.ShmOffs))
			b := rpcc.shmemSegment.GetBuf()
			for i := range rep.ShmOffs {
				start := rep.ShmOffs[i]
				end := start + rep.ShmLens[i]
				// The offsets come from another process; a frame that doesn't lie
				// inside this segment would otherwise panic on the slice, or
				// silently hand the caller unrelated bytes to unmarshal.
				if end < start || end > uint64(len(b)) {
					return 0, serr.NewErr(serr.TErrUnreachable, fmt.Sprintf("shmem delegated RPC(%v) frame %d of %d out of segment: [%v,%v) size %v", rpcIdx, i, len(rep.ShmOffs), start, end, len(b)))
				}
				frames[i] = sessp.NewFrame(b[start:end], nil)
			}
			outiov.AppendFrames(frames)
			rep.Blob.SetIoVec(outiov)
		} else {
			db.DFatalf("Err: rpcclnt using shared memory but reply isn't using shared memory")
		}
	}
	transferDur := time.Since(rep.TransferStartPB.AsTime())
	perf.LogSpawnLatency("DelegatedRPC.RunRPC %d", sp.NOT_SET, perf.TIME_NOT_SET, start, rpcIdx)
	start = time.Now()
	defer func(start time.Time) {
		perf.LogSpawnLatency("DelegatedRPC.Unmarshal %d", sp.NOT_SET, perf.TIME_NOT_SET, start, rpcIdx)
	}(start)
	return transferDur, processWrappedRPCRep(rep.Blob.GetIoVec(), res, outblob)
}

// Fetch a batch of delegated RPC results
func (rpcc *RPCClnt) BatchFetchDelegatedRPCs(idxs []uint64, nIOV int) error {
	req := &spproxyproto.SigmaMultiDelegatedRPCReq{
		RPCIdxs: idxs,
	}
	multiRep := &spproxyproto.SigmaMultiDelegatedRPCRep{
		Blob: &rpcproto.Blob{
			Iov: make([][]byte, nIOV),
		},
	}
	err := rpcc.rpc(true, "SPProxySrvAPI.GetMultiDelegatedRPCReplies", req, multiRep)
	if err != nil {
		return err
	}
	start := 0
	for i, rpcIdx := range idxs {
		rpcc.rc.Register(rpcIdx)
		end := start + int(multiRep.NIOVs[i])
		rep := &spproxyproto.SigmaDelegatedRPCRep{
			Blob: &rpcproto.Blob{
				Iov: multiRep.Blob.Iov[start:end],
			},
			Err: multiRep.Errs[i],
		}
		rpcc.rc.Put(rpcIdx, rep, err)
		start = end
	}
	return nil
}

func (rpcc *RPCClnt) StatsClnt() map[string]*rpc.MethodStatSnapshot {
	return rpcc.si.Stats()
}

func (rpcc *RPCClnt) StatsSrv() (*rpc.RPCStatsSnapshot, error) {
	return rpcc.ch.StatsSrv()
}

func (rpcc *RPCClnt) Channel() channel.RPCChannel {
	return rpcc.ch
}

File: cpp/rpc/clnt.cc
#include <rpc/blob.h>
#include <rpc/clnt.h>
#include <util/perf/perf.h>

namespace sigmaos {
namespace rpc {

bool Clnt::_l = sigmaos::util::log::init_logger(RPCCLNT);
bool Clnt::_l_e = sigmaos::util::log::init_logger(RPCCLNT_ERR);

std::expected<int, sigmaos::serr::Error> Clnt::BatchFetchDelegatedRPCs(
    std::vector<uint64_t> &rpc_idxs, int n_iov) {
  log(RPCCLNT, "BatchFetchDelegatedRPCs {}", (int)rpc_idxs.size());
  auto out_iov = new sigmaos::io::iovec::IOVec();
  out_iov->AddBuffers(n_iov);
  SigmaMultiDelegatedRPCReq req;
  for (uint64_t rpc_idx : rpc_idxs) {
    req.add_rpcidxs(rpc_idx);
  }
  SigmaMultiDelegatedRPCRep rep;
  Blob blob;
  auto iov = blob.mutable_iov();
  for (int i = 0; i < n_iov; i++) {
    iov->AddAllocated(out_iov->GetBuffer(i)->Get());
  }
  rep.set_allocated_blob(&blob);
  {
    // If there was no cached reply, run the delegated RPC
    auto res = rpc(true, "SPProxySrvAPI.GetMultiDelegatedRPCReplies", req, rep);
    if (!res.has_value()) {
      return res;
    }
  }
  int start = 0;
  for (int i = 0; i < rpc_idxs.size(); i++) {
    int end = start + rep.niovs(i);
    uint64_t rpc_idx = rpc_idxs[i];
    _cache.Register(rpc_idx);
    auto delegated_rep = std::make_shared<SigmaDelegatedRPCRep>();
    auto blob = new Blob();
    auto iov = blob->mutable_iov();
    for (int j = start; j < end; j++) {
      iov->AddAllocated(out_iov->GetBuffer(j)->Get());
    }
    delegated_rep->set_allocated_blob(blob);
    *delegated_rep->mutable_err() = rep.errs(i);
    // Cache the result
    _cache.Put(rpc_idx, delegated_rep, nullptr);
    start = end;
  }
  return 0;
}

// Retrieve the result of a delegated RPC
std::expected<int, sigmaos::serr::Error> Clnt::DelegatedRPC(
    uint64_t rpc_idx, google::protobuf::Message &delegated_rep,
    std::shared_ptr<std::vector<std::shared_ptr<std::string_view>>> views) {
  log(RPCCLNT, "DelegatedRPC {}", (int)rpc_idx);
  // Sanity check
  if ((_shmem && !views) || (!_shmem && views)) {
    fatal("Try delegated RPC with mismatching shmem & views");
  }
  auto out_iov = std::make_shared<sigmaos::io::iovec::IOVec>();
  if (!_shmem) {
    // Prepend empty slots to the out iovec for the marshaled delegated reply
    // and its RPC wrapper
    out_iov->AddBuffers(2);
  }
  // Extract any output IOVecs from the delegated reply RPC
  extract_blob_iov(delegated_rep, out_iov);
  // Create the delegated request
  SigmaDelegatedRPCReq req;
  req.set_rpcidx(rpc_idx);
  req.set_useshmem(_shmem != nullptr);
  auto rep = std::make_shared<SigmaDelegatedRPCRep>();
  Blob blob;
  // Set IOVec buffers if not using shared memory
  auto iov = blob.mutable_iov();
  if (!_shmem) {
    // Add the output buffers for the RPC wrapper & serialized RPC reply, as
    // well as the delegated reply's blob output buffers to the RPC's blob
    for (int i = 0; i < out_iov->Size(); i++) {
      iov->AddAllocated(out_iov->GetBuffer(i)->Get());
    }
  }
  rep->set_allocated_blob(&blob);
  bool reply_cached = false;
  // TODO: remove TS
  auto start = GetCurrentTime();
  {
    auto res = _cache.Get(rpc_idx, out_iov);
    if (!res.has_value()) {
      return res;
    }
    reply_cached = res.value();
  }
  if (reply_cached) {
    log(RPCCLNT, "DelegatedRPC({}) reply cached", (int)rpc_idx);
  } else {
    log(RPCCLNT, "DelegatedRPC({}) reply not cached", (int)rpc_idx);
    // If there was no cached reply, run the delegated RPC
    auto res = rpc(true, "SPProxySrvAPI.GetDelegatedRPCReply", req, *rep);
    if (!res.has_value()) {
      return res;
    }
  }
  // If using shared memory
  if (_shmem) {
    if (rep->useshmem()) {
      // Sanity check that enough buffers were supplied
      if (views->size() != (rep->shmoffs().size() - 2)) {
        fatal("Wrong num buffers supplied for shared-memory delRPC: {} != {}",
              views->size(), rep->shmoffs().size());
      }
      log(RPCCLNT, "DelegatedRPC({}) using shared memory", (int)rpc_idx);
      // Set the output buffers for the RPC wrapper & serialized RPC reply
      for (int i = 0; i < 2; i++) {
        uint64_t off = rep->shmoffs(i);
        size_t len = (size_t)rep->shmlens(i);
        auto b =
            std::make_shared<std::string>((char *)_shmem->GetBuf() + off, len);
        out_iov->AppendBuffer(std::make_shared<sigmaos::io::iovec::Buffer>(b));
      }
      // Skip the first two IOVs since those are the RPC wrapper & RPC struct
      for (int i = 2; i < rep->shmoffs().size(); i++) {
        uint64_t off = rep->shmoffs(i);
        size_t len = (size_t)rep->shmlens(i);
        views->at(i - 2) = std::make_shared<std::string_view>(
            (char *)_shmem->GetBuf() + off, len);
      }
      log(RPCCLNT, "DelegatedRPC({}) using shared memory set views successful",
          (int)rpc_idx);
    }
  }
  // Process the delegated, wrapped RPC reply
  start = GetCurrentTime();
  auto res = process_wrapped_reply(rpc_idx, out_iov, delegated_rep);
  log(ALWAYS, "DelegatedRPC({}) process wrapped reply latency {}ms",
      (int)rpc_idx, LatencyMS(start));
  log(RPCCLNT, "DelegatedRPC done {}", (int)rpc_idx);
  {
    // Release ownership of blob (which is a local stack variable declared
    // above)
    auto blob = rep->release_blob();
  }
  return res;
}

// Perform an RPC
std::expected<int, sigmaos::serr::Error> Clnt::RPC(
    std::string method, google::protobuf::Message &req,
    google::protobuf::Message &rep) {
  {
    auto res = check_channel_init();
    if (!res.has_value()) {
      return res;
    }
  }
  return rpc(false, method, req, rep);
}

std::expected<int, sigmaos::serr::Error> Clnt::check_channel_init() {
  // Fast-path: check if channel is initialized
  if (_chan->IsInitialized()) {
    return 0;
  }

  std::lock_guard<std::mutex> guard(_mu);
  // Slow-path: check again, now holding the lock
  if (_chan->IsInitialized()) {
    return 0;
  }
  // Initialize the channel
  auto res = _chan->Init();
  if (!res.has_value()) {
    log(RPCCLNT_ERR, "Error initialize channel: {}", res.error().String());
    return res;
  }
  return 0;
}

std::expected<int, sigmaos::serr::Error> Clnt::rpc(
    bool delegate, std::string method, google::protobuf::Message &req,
    google::protobuf::Message &rep) {
  // Create an IOV for RPC inputs
  auto in_iov = std::make_shared<sigmaos::io::iovec::IOVec>();
  auto req_data = std::make_shared<std::string>();
  // Serialize the request and append it to the IOV
  req.SerializeToString(req_data.get());
  in_iov->AppendBuffer(std::make_shared<sigmaos::io::iovec::Buffer>(req_data));
  // Extract any input IOVecs from the request RPC
  extract_blob_iov(req, in_iov);
  auto out_iov = std::make_shared<sigmaos::io::iovec::IOVec>();
  // Prepend 2 empty slots to the out iovec: one for the rpcproto.Rep
  // wrapper, and one for the marshaled res proto.Message
  out_iov->AddBuffers(2);
  // Extract any output IOVecs from the response RPC
  extract_blob_iov(rep, out_iov);
  log(RPCCLNT, "out_iov len {}", out_iov->Size());
  uint64_t seqno = _seqno++;
  // Wrap the RPC and execute it
  auto res = wrap_and_run_rpc(delegate, seqno, method, in_iov, out_iov);
  if (!res.has_value()) {
    return std::unexpected(res.error());
  }
  // Process the wrapped response
  return process_wrapped_reply(seqno, out_iov, rep);
}

std::expected<int, sigmaos::serr::Error> Clnt::wrap_and_run_rpc(
    bool delegate, uint64_t seqno, std::string method,
    const std::shared_ptr<sigmaos::io::iovec::IOVec> in_iov,
    std::shared_ptr<sigmaos::io::iovec::IOVec> out_iov) {
  auto wrapped_in_iov = std::make_shared<sigmaos::io::iovec::IOVec>();
  Req req;
  auto wrapper_req_data = std::make_shared<std::string>();
  req.set_method(method);
  // Wrap the request, serialize the wrapper, and write it together with the
  // request data.
  req.SerializeToString(wrapper_req_data.get());
  wrapped_in_iov->AppendBuffer(
      std::make_shared<sigmaos::io::iovec::Buffer>(wrapper_req_data));
  wrapped_in_iov->Resize(in_iov->Size() + 1);
  for (int i = 0; i < in_iov->Size(); i++) {
    wrapped_in_iov->SetBuffer(i + 1, in_iov->GetBuffer(i));
  }

  // Create the call object to be sent, and perform the RPC.
  auto wrapped_call =
      std::make_shared<io::transport::Call>(seqno, wrapped_in_iov, out_iov);
  auto start = GetCurrentTime();
  std::expected<std::shared_ptr<sigmaos::io::transport::Call>,
                sigmaos::serr::Error>
      res;
  if (delegate) {
    // Sanity check
    if (!_delegate_chan) {
      fatal("Try to run a delegated RPC with a null delegate channel");
    }
    res = _delegate_chan->SendReceive(wrapped_call);
  } else {
    res = _chan->SendReceive(wrapped_call);
  }
  if (!res.has_value()) {
    log(RPCCLNT_ERR, "Error sendreceive: {}", res.error().String());
    return std::unexpected(res.error());
  }
  log(PROXY_RPC_LAT, "RPCClnt SendReceive");
  return 0;
}

std::expected<int, sigmaos::serr::Error> Clnt::process_wrapped_reply(
    uint64_t seqno, std::shared_ptr<sigmaos::io::iovec::IOVec> out_iov,
    google::protobuf::Message &rep) {
  log(RPCCLNT, "Deserialize wrapper for reply seqno: {}", seqno);
  Rep wrapped_rep;
  // Deserialize the wrapper
  auto wrapper_rep_buf = out_iov->GetBuffer(0);
  auto start = GetCurrentTime();
  wrapped_rep.ParseFromString(*wrapper_rep_buf->Get());
  log(RPCCLNT, "Remove wrapper buffer for reply seqno: {}", seqno);
  log(PROXY_RPC_LAT, "RPCClnt Parse wrapper");
  // Remove the wrapper from the out IOVec
  out_iov->RemoveBuffer(0);
  log(RPCCLNT, "Done remove wrapper buffer for reply seqno: {}", seqno);
  // TODO: Record stats
  //	rpcc.si.Stat(method, time.Since(start).Microseconds())
  // If there was an error in the RPC stack, bail out
  if (wrapped_rep.err().errcode() != sigmaos::serr::TErrNoError) {
    log(RPCCLNT_ERR, "error in RPC stack");
    log(RPCCLNT_ERR, "error in RPC stack error:{}",
        wrapped_rep.err().ShortDebugString());
    return std::unexpected(sigmaos::serr::Error(
        sigmaos::serr::TErrUnreachable,
        std::format("rpc error: {}", wrapped_rep.err().ShortDebugString())));
  }
  log(RPCCLNT, "Deserialize reply data");
  // Deserialize the reply
  auto rep_data_buf = out_iov->GetBuffer(0);
  start = GetCurrentTime();
  rep.ParseFromString(*rep_data_buf->Get());
  log(PROXY_RPC_LAT, "RPCClnt Parse reply lat:{}ms", (int)LatencyMS(start));
  log(RPCCLNT, "Remove reply data buffer");
  // Remove the first element in iov, which contains the serialized reply
  // message
  out_iov->RemoveBuffer(0);
  log(RPCCLNT, "Set blob IOV len {}", out_iov->Size());
  // Set the reply's blob's IOV to point to the returned data, if applicable.
  set_blob_iov(out_iov, rep);
  log(RPCCLNT, "Done set blob IOV");
  return 0;
}

};  // namespace rpc
};  // namespace sigmaos
